MapleServer2/PacketHandlers/Game/MasteryHandler.cs

```diff
@@ -0,0 +1,194 @@
+using MaplePacketLib2.Tools;
+using MapleServer2.Constants;
+using MapleServer2.Data.Static;
+using MapleServer2.Enums;
+using MapleServer2.Packets;
+using MapleServer2.Servers.Game;
+using MapleServer2.Tools;
+using MapleServer2.Types;
+using Microsoft.Extensions.Logging;
+
+namespace MapleServer2.PacketHandlers.Game
+{
+    public class MasteryHandler : GamePacketHandler
+    {
+        public override RecvOp OpCode => RecvOp.CONSTRUCT_RECIPE;
+
+        public MasteryHandler(ILogger<MasteryHandler> logger) : base(logger) { }
+
+        private enum MasteryMode : byte
+        {
+            RewardBox = 0x01,
+            CraftItem = 0x02
+        }
+
+        public override void Handle(GameSession session, PacketReader packet)
+        {
+            MasteryMode mode = (MasteryMode) packet.ReadByte();
+            switch (mode)
+            {
+                case MasteryMode.RewardBox:
+                    HandleRewardBox(session, packet);
+                    break;
+                case MasteryMode.CraftItem:
+                    HandleCraftItem(session, packet);
+                    break;
+                default:
+                    IPacketHandler<GameSession>.LogUnknownMode(mode);
+                    break;
+            }
+        }
+
+        private void HandleRewardBox(GameSession session, PacketReader packet)
+        {
+            int rewardBoxDetails = packet.ReadInt();
+            int type = rewardBoxDetails / 1000;
+            int grade = rewardBoxDetails % 100;
+
+            int? rewardBoxItemId = MasteryMetadataStorage.GetMasteryTierRewardItemId(type, grade);
+            if (rewardBoxItemId == null)
+            {
+                Logger.LogError(
+                    $"Unknown reward box item ID of mastery type: {type}, tier: {grade} from user: {session.Player.Name}");
+                return;
+            }
+
+            Logger.LogError(
+                $"reward box item ID: {rewardBoxItemId}, mastery type: {type}, tier: {grade} from user: {session.Player.Name}");
+
+            Item rewardBox = new Item((int) rewardBoxItemId) {Amount = 1};
+
+            // give player the reward box item
+            InventoryController.Add(session, rewardBox, true);
+
+            // mark reward box as claimed
+            session.Send(MasteryPacket.ClaimReward(rewardBoxDetails, 1, (int) rewardBoxItemId));
+        }
+
+        private void HandleCraftItem(GameSession session, PacketReader packet)
+        {
+            int recipeId = packet.ReadInt();
+
+            // attempt to oad the recipe metadata
+            RecipeMetadata recipe = RecipeMetadataStorage.GetRecipe(recipeId);
+            if (recipe == null)
+            {
+                Logger.LogError($"Unknown recipe ID {recipeId} from user: {session.Player.Name}");
+                return;
+            }
+            
+            // does the play have enough mesos for this recipe?
+            if (!PlayerHasEnoughMesos(session, recipe))
```

Comment:
You can just used Wallet Modify function here, it returns a bool and if true also changes the meso amount!